openedx/core/djangoapps/programs/utils.py

```diff
@@ -127,6 +127,43 @@ def engaged_programs(self):
 
         return programs
 
+    @cached_property
+    def failed_course_runs(self):
+        """
+        Determine which course runs have been failed by the user.
+
+        Returns:
+            list of strings, each representing a certificate course run key
```

Comment:
Definitely refer to them as course runs, sorry I wasn't clear. I'm just looking for you to be careful about use of the word `key`. In edx-platform, that often means a `CourseKey`, whereas ID means a string.